### DISCUSSION TITLE
Camel Karavan : Servlet as component to rest configuration.

### DISCUSSION THREAD

**Discussion by MN-Dheeraj** [apache/camel-karavan#1240] [Q&A] [answered]
### Ask a question

I am adding my spring boot code to karavan to build route of an API. In my code, there is a common class for route builder that has this line.
restConfiguration().component("servlet");

I see that servlet is being added as component to the rest configuration. So, I tried to do that on karavan by going to REST>Create configuration and selected the component as servlet from the dropdown.
I get this error however.

Error:
--------
2024-04-23 04:23:25.089  WARN 15 --- [ ReloadOnDemand] support.RouteOnDemandReloadStrategy : Error reloading routes due to Failed to create route route-a27c: Route(route-a27c)[From[rest:post:/LoanAccountDetails] -> [pr... because of Failed to resolve endpoint: rest://post:/LoanAccountDetails due to: No RestConfiguration for component: platform-http found, RestConfiguration targets: servlet. This exception is ignored.
org.apache.camel.FailedToCreateRouteException: Failed to create route route-a27c: Route(route-a27c)[From[rest:post:/LoanAccountDetails] -> [pr... because of Failed to resolve endpoint: rest://post:/LoanAccountDetails due to: No RestConfiguration for component: platform-http found, RestConfiguration targets: servlet
	at org.apache.camel.reifier.RouteReifier.createRoute(RouteReifier.java:85) ~[camel-core-reifier-4.4.0.jar:4.4.0]
	at org.apache.camel.impl.DefaultModelReifierFactory.createRoute(DefaultModelReifierFactory.java:49) ~[camel-core-engine-4.4.0.jar:4.4.0]
	at org.apache.camel.impl.DefaultCamelContext.startRouteDefinitions(DefaultCamelContext.java:728) ~[camel-core-engine-4.4.0.jar:4.4.0]
	at org.apache.camel.impl.DefaultModel.addRouteDefinitions(DefaultModel.java:249) ~[camel-core-engine-4.4.0.jar:4.4.0]

Here is my route on karavan. 
- route:
    id: route-a27c
    nodePrefixId: route-06e
    from:
      id: from-1df0
      uri: rest
      parameters:
        method: post
        path: /LoanAccountDetails
      steps:
        - process:
            id: process-d06b
            description: Interception and logging to DB
            ref: Esbloginterceptorprocessor
------other steps----------

Can someone please guide me in adding servlet as component? I tried adding 'Consumer Component Name' of my 'from' REST component on my route itself. But, it doesn't work. When I make my API call, I get 404. Without, this modification, it works fine.

### Variant

Web Application

### Container Management (if applicable)

Docker

### Operating System (if applicable)

Windows

### Version

4.4.0

**Comment by davsclaus** [comment]
do not use servlet, you should use platform-http which is the default, so you do not have to configure it.

**Reply by MN-Dheeraj** [reply]
Hi @davsclaus ,

Do you mean to say that adding servlet as component doesn't work at all?
Actually, In my camel-springboot code, this happens

HttpServletRequest req = exchange.getIn(HttpServletRequest.class);
String remoteAddr = null;
if (req != null) {
	remoteAddr = req.getRemoteAddr();
	String method = Exchange.HTTP_METHOD;
	apiLogDetail.setServiceName(apiName);
        apiLogDetail.setEndPoint(req.getRequestURL().toString());
	//saving apiLogDetail to DB			
}

I am getting req as null probably because I commented this line because of the above mentioned issue.
restConfiguration().component("servlet");

When you say 'don't use servlet', do you suggest any alternative or that there is no provision to use it on karavan at all?

**Comment by davsclaus** [comment]
karavan comes with platform-http as its supported http server, so use that

**Reply by MN-Dheeraj** [reply]
This is my code.

HttpServletRequest req = exchange.getIn(HttpServletRequest.class);
remoteAddr = req.getRemoteAddr();
String url = req.getRequestURL().toString();

If servlet can't be used, the request would obviously be null.
Can you suggest an alternative logic in case of platform-http?

My code is in spring boot.

**Reply by MN-Dheeraj** [reply]
@davsclaus ,

I see that I can create rest configuration on karavan. In the list of the dropdown on the component, I see 'servlet' also. May I know how this works?

<img width="735" alt="servlet_component_karavan" src="https://github.com/apache/camel-karavan/assets/160023109/016888a9-f24a-45c2-985e-e0fc0b7691ec">

**Reply by davsclaus** [reply, marked as answer]
Karavan ui is ALL of camel and shows everything - also older http servers like jetty and coap for mqtt protocols. 
But when you run in karavan web it runs only in vanilla camel standalone, and platform-http is only supported there (auto detected so you dont need to select anyway).

**Reply by MN-Dheeraj** [reply]
I have this in Vscode as well. Is it not possible there as well?

Also, I will mark it as answer myself, once it is fully clarified.
